### PR TITLE
Rework processing of composition events

### PR DESCRIPTION
**Pull request by alexanderPolosatov**

### CODE REVIEW

File: blots/block.js [hunk 1]
```diff
@@ -77,7 +77,10 @@ class Block extends BlockBlot {
 
   insertBefore(blot, ref) {
     const { head } = this.children;
-    super.insertBefore(blot, ref);
+
+    if (!ref || this.domNode.contains(ref.domNode)) {
+      super.insertBefore(blot, ref);
+    }
```
Comment (ksercs) [inline blots/block.js:83]: could u please clarify this change?

File: core/quill.js [hunk 2]
```diff
@@ -64,6 +65,7 @@ class Quill {
   }
 
   constructor(container, options = {}) {
+    this.formattingStarted = false;
```
Comment (ksercs) [inline core/quill.js:68]: It looks like some debug row forgotten to be deleted :)

File: core/quill.js [hunk 3]
```diff
@@ -445,6 +449,10 @@ class Quill {
       true,
     );
   }
+
+  startFormat() {
```
Comment (ksercs) [inline core/quill.js:453]: I shared my thoughts in the related PR

File: core/selection.js [hunk 4]
```diff
@@ -14,21 +14,20 @@ class Range {
...
     this.cursor = this.scroll.create('cursor', this);
     // savedRange is last non-null range
     this.savedRange = new Range(0, 0);
     this.lastRange = this.savedRange;
     this.lastNative = null;
-    this.handleComposition();
     this.handleDragging();
     this.emitter.listenDOM('selectionchange', document, () => {
-      if (!this.mouseDown && !this.composing) {
+      if (!this.mouseDown && !this.composition.isComposing) {
```
Comment (ksercs) [inline core/selection.js:30]: let's add a "public" getter for it?

File: core/selection.js [hunk 5]
```diff
@@ -355,6 +335,9 @@ class Selection {
           endNode = endNode.parentNode;
         }
         const range = document.createRange();
+        if (startOffset < 0) {
+          startOffset = 0;
+        }
```
Comment (ksercs) [inline core/selection.js:340]: Hm, it's just interesting for me - did u find a real scenario showing the previous code was broken?

File: core/selection.js [hunk 6]
```diff
@@ -355,6 +335,9 @@ class Selection {
           endNode = endNode.parentNode;
         }
         const range = document.createRange();
+        if (startOffset < 0) {
```
Comment (ksercs) [inline core/selection.js:338]: we can create an util function `toNotNegative` to reduce the copypaste a bit

File: test/unit/core/input.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import Input from '../../../modules/input';
+
+describe('subscriptions', function () {
+  it('subscribtion on beforeinput event should be', function () {
```
Comment (ksercs) [inline test/unit/core/input.js:4]:
```suggestion
  it('subscription on beforeinput event should be', function () {
```

File: test/unit/core/input.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    };
+
+    // eslint-disable-next-line no-new
+    new Input(quill, {});
+
+    expect(addEventListenerCallCount).toEqual(1);
+    expect(eventName).toEqual('beforeinput');
+    expect(typeof handler).toEqual('function');
+  });
+});
```
Comment (ksercs) [inline test/unit/core/input.js:26]: I see some kind of an "integration" test in `quill.js` for `beforeinput` event but don't see a similar one for `compositionStart`. Cannot we add it?

File: test/unit/core/composition.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+      this.scrollMock = getScrollMock(this.target);
+
+      this.composition = new Composition(
+        this.scrollMock.scroll,
+        this.emitMock.emitter,
+      );
+    });
+
+    it('trigger compositionstart event', function () {
```
Comment (ksercs) [inline test/unit/core/composition.js:92]: why don't i wanna use our usual approach for test naming? I mean answering on a question "how should it work"

File: core/selection.js (hunk 5 above)
Comment (ksercs) [inline core/selection.js:340]: What about tests for these changes in the selection module? :)

File: core/quill.js (hunk 2 above)
Comment (alexanderPolosatov) [reply core/quill.js:68]: yes. it's from prev changes. will fixed

File: core/selection.js (hunk 4 above)
Comment (alexanderPolosatov) [reply core/selection.js]: added

File: core/quill.js (hunk 3 above)
Comment (alexanderPolosatov) [reply core/quill.js]: refactored. at present, some additional calls special methods not needed

File: core/composition.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    });
+
+    scroll.domNode.addEventListener('compositionend', (event) => {
+      if (this.isComposing) {
+        this.handleCompositionEnd(event);
+      }
+    });
+  }
+
+  isConposingStagted() {
```
Comment (ksercs) [inline core/composition.js:22]: something went wrong here :) 
Maybe we can just name internal prop as `_isComposing` and rename this method with `isComposing`?

File: test/unit/core/quill.js [hunk 11]
```diff
@@ -216,6 +216,17 @@ describe('Quill', function () {
 
       expect(this.quill.getFormat()).toEqual({});
     });
+
+    it('compotion.handleCompositionEnd method should be called on format', function () {
```
Comment (ksercs) [inline test/unit/core/quill.js:220]:
```suggestion
    it('composition.handleCompositionEnd method should be called on format', function () {
```

File: test/unit/core/selection.js [hunk 12]
```diff
@@ -77,6 +77,14 @@ describe('Selection', function () {
       expect(range.length).toEqual(0);
     });
 
+    it('should no exceptions when offset is negative', function () {
```
Comment (ksercs) [inline test/unit/core/selection.js:80]:
```suggestion
    it('no exceptions should be presented when offset is negative', function () {
```

File: core/composition.js (hunk 10 above)
Comment (alexanderPolosatov) [reply core/composition.js:22]: we can't do it without changing of lint rules. At present we have enabled 'no-underscore-dangle'. I'm not sure what we should change this rule...
BTW, method has typo. I will fix it anyway.

Comment (ksercs) [reply core/composition.js:22]: then i suggest `isCompositionInProgress`

Comment (alexanderPolosatov) [reply core/composition.js]: renamed

**Review by ksercs — APPROVED** [review]

**Pull request merged**